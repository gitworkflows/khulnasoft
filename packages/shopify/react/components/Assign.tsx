import { KhulnasoftStore, onChange, withKhulnasoft } from '@khulnasoft.com/react';
import * as React from 'react';
import { AssignBlockProps } from '../interfaces/component-props';

export class AssignBlock extends React.Component<AssignBlockProps> {
  ran = false;

  constructor(props: AssignBlockProps) {
    super(props);
  }

  run() {
    const { expression, khulnasoftState } = this.props;

    if (expression && khulnasoftState) {
      if (khulnasoftState.context.shopify) {
        khulnasoftState.context.shopify.liquid.assign(expression, onChange.target(khulnasoftState.state));
        this.ran = true;
        return true;
      }
    }
    return false;
  }

  render() {
    this.run();
    return null;
  }
}

withKhulnasoft(AssignBlock, {
  name: 'Shopify:Assign',
  hideFromInsertMenu: true,
  noWrap: true,
  inputs: [
    {
      name: 'expression',
      type: 'javascript',
      // TODO: type: 'code', language: 'javascript'
    },
  ],
});
