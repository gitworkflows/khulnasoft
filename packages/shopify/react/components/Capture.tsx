import { KhulnasoftStore, onChange, withKhulnasoft } from '@khulnasoft.com/react';
import * as React from 'react';
import { CaptureBlockProps } from '../interfaces/component-props';

export class CaptureBlock extends React.Component<CaptureBlockProps> {
  ran = false;
  constructor(props: CaptureBlockProps) {
    super(props);
  }

  run() {
    const { expression, variableName, khulnasoftState } = this.props;

    if (expression && khulnasoftState?.context?.shopify) {
      onChange.target(khulnasoftState.state)[variableName] =
        khulnasoftState.context.shopify.liquid.render(expression, khulnasoftState.state);

      this.ran = true;
      return true;
    }

    return false;
  }

  render() {
    this.run();
    return null;
  }
}

withKhulnasoft(CaptureBlock, {
  name: 'Shopify:Capture',
  hideFromInsertMenu: true,
  noWrap: true,
  inputs: [
    {
      name: 'expression',
      type: 'javascript',
    },
  ],
});
