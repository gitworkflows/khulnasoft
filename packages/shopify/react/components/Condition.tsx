import khulnasoft, {
  Khulnasoft,
  KhulnasoftBlockComponent,
  KhulnasoftElement,
  KhulnasoftStore,
  stringToFunction,
  withKhulnasoft,
} from '@khulnasoft.com/react';
import * as React from 'react';
import { ConditionBlockProps } from '../interfaces/component-props';

export class ConditionBlock extends React.Component<ConditionBlockProps> {
  private getMatchingBranch(branches = this.props.branches) {
    if (!(branches && branches.length)) {
      return null;
    }

    for (let i = 0; i < branches.length; i++) {
      const branch = branches[i];
      const isLast = i === branches.length - 1;
      if (isLast && !branch.expression) {
        return { index: i, blocks: branch.blocks };
      }
      const fn = stringToFunction(branch.expression || '');
      const result = fn(
        this.props.khulnasoftState && this.props.khulnasoftState.state,
        null,
        this.props.khulnasoftBlock,
        khulnasoft,
        null,
        this.props.khulnasoftState!.update,
        Khulnasoft,
        this.props.khulnasoftState?.context
      );
      if (result) {
        return { index: i, blocks: branch.blocks };
      }
    }
    return null;
  }

  render() {
    const result = this.getMatchingBranch();
    if (!result) {
      return null;
    }

    return result.blocks.map(block => <KhulnasoftBlockComponent key={block.id} block={block} />);
  }
}

const defaultBlock = {
  '@type': '@khulnasoft.com/sdk:Element',
  responsiveStyles: {
    large: {
      display: 'flex',
      flexDirection: 'column',
      alignItems: 'stretch',
      flexShrink: '0',
      position: 'relative',
      marginTop: '30px',
      textAlign: 'center',
      lineHeight: 'normal',
      height: 'auto',
    },
  },
  component: {
    name: 'Text',
    options: {
      text: '<p>Enter some text...</p>',
    },
  },
};

withKhulnasoft(ConditionBlock, {
  name: 'Shopify:Condition',
  hideFromInsertMenu: true,
  noWrap: true,
  inputs: [
    {
      name: 'branches',
      type: 'array',
      subFields: [
        {
          name: 'expression',
          type: 'javascript',
        },
        {
          name: 'blocks',
          type: 'array',
          hideFromUI: true,
          defaultValue: [defaultBlock],
        },
      ],
    },
  ],
});
