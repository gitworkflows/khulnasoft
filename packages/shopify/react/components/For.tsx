import {
  KhulnasoftStore,
  withKhulnasoft,
  KhulnasoftStoreContext,
  KhulnasoftBlockComponent,
  KhulnasoftElement,
  Khulnasoft,
} from '@khulnasoft.com/react';
import React, { useState } from 'react';

interface ForBlockProps {
  repeat: {
    itemName: string;
    collection: string;
  };
  khulnasoftState?: KhulnasoftStore;
  khulnasoftBlock?: KhulnasoftElement;
}

export function ForBlock(props: ForBlockProps) {
  let arrayItems: any[] = [];
  const { repeat, khulnasoftState } = props;

  const collectionPath = repeat.collection;
  const collectionName = (collectionPath || '').trim().split('(')[0].trim().split('.').slice(-1)[0];

  const itemName = repeat.itemName || (collectionName ? collectionName + 'Item' : 'item');

  if (itemName && collectionName && khulnasoftState && Khulnasoft.isBrowser) {
    if (khulnasoftState.context.shopify) {
      const renderedItems = khulnasoftState.context.shopify.liquid.get(
        collectionPath,
        khulnasoftState.state
      );

      if (Array.isArray(renderedItems)) {
        arrayItems = renderedItems;
      }
    }
  }

  return arrayItems?.map((item, index) => {
    const scopedState: any = {
      forloop: {
        index0: index,
        index: index + 1,
        first: index === 0,
        last: index === arrayItems.length - 1,
        length: arrayItems.length,
        rindex: arrayItems.length - index,
        rindex0: arrayItems.length - index - 1,
      },
    };

    // add the itemName (ie "product") to the state so the children can access
    scopedState[itemName] = item;

    return (
      <KhulnasoftStoreContext.Consumer key={index}>
        {store => (
          <KhulnasoftStoreContext.Provider
            value={{
              ...store,
              state: {
                ...store.state,
                ...scopedState,
              },
            }}
          >
            {props.khulnasoftBlock?.children?.map(item => (
              <KhulnasoftBlockComponent block={item} key={item.id} />
            ))}
          </KhulnasoftStoreContext.Provider>
        )}
      </KhulnasoftStoreContext.Consumer>
    );
  });
}

withKhulnasoft(ForBlock, {
  name: 'Shopify:For',
  hideFromInsertMenu: true,
  noWrap: true,
  inputs: [
    {
      name: 'expression',
      type: 'javascript',
    },
  ],
});
