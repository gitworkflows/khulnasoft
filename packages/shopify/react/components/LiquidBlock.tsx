import React, { useState, useEffect, useRef } from 'react';
import { Khulnasoft, KhulnasoftElement, KhulnasoftStore, khulnasoft } from '@khulnasoft.com/react';
import { findAndRunScripts } from '../functions/find-and-run-scripts';

interface LiquidBlockProps {
  options?: Record<string, number | boolean | string>;
  templatePath?: string;
  khulnasoftBlock?: KhulnasoftElement;
  khulnasoftState?: KhulnasoftStore;
}
const cache: { [key: string]: string } = {};

const refs =
  Khulnasoft.isBrowser &&
  Array.from(document.querySelectorAll('[khulnasoft-liquid-block]')).reduce((memo, item) => {
    memo[item.getAttribute('khulnasoft-liquid-block')!] = item;
    return memo;
  }, {} as { [key: string]: Element });

export const LiquidBlock = ({
  options,
  khulnasoftBlock,
  templatePath,
  khulnasoftState,
}: LiquidBlockProps) => {
  const [html, setHtml] = useState<string | null>(null);
  const [loading, setLoading] = useState(0);
  const ref = useRef<HTMLDivElement | null>(null);
  const blockName = templatePath?.split('/')[1].replace(/\.liquid$/, '')!;
  const renderArgs = JSON.stringify(options);

  useEffect(() => {
    const blockId = khulnasoftBlock?.id;
    const node = blockId && refs && refs[blockId];

    const cacheKey = blockName + renderArgs;

    if (!node || Khulnasoft.isEditing || Khulnasoft.isPreviewing) {
      if (cache[cacheKey]) {
        setHtml(cache[cacheKey]);
        return;
      }

      if (!blockName) {
        return;
      }

      if (cache[cacheKey]) {
        setHtml(cache[cacheKey]);
        return;
      }

      if (Khulnasoft.isEditing) {
        setLoading(loading => loading + 1);
      }

      const previewThemeID =
        (Khulnasoft.isBrowser && (window as any).Shopify?.theme?.id) ||
        khulnasoftState?.state?.location?.query?.preview_theme_id;

      fetch(
        `${khulnasoft.host}/api/v1/shopify/data/render-liquid-snippet?snippet=${blockName}&apiKey=${
          khulnasoftState?.context.apiKey
        }&args=${encodeURIComponent(renderArgs)}${
          previewThemeID ? `&preview_theme_id=${previewThemeID}` : ''
        }${Khulnasoft.isEditing ? `&cachebust=true` : ''}`
      )
        .then(res => res.json())
        .then(json => {
          const text = json.liquidSnippet;
          cache[cacheKey] = text;
          setHtml(text);
          setTimeout(() => {
            if (ref.current) {
              findAndRunScripts(ref.current);
            }
          }, 10);
        })
        .finally(() => {
          if (Khulnasoft.isEditing) {
            setLoading(loading => Math.max(loading - 1, 0));
          }
        });
    } else {
      if (node && ref.current) {
        ref.current.parentNode?.replaceChild(node, ref.current);
      }
    }
  }, [options, blockName]);

  return (
    <>
      {Boolean(loading) && (
        <>
          <div
            style={{
              pointerEvents: 'none',
              zIndex: 10,
              height: 50,
              width: 50,
              alignItems: 'center',
              justifyContent: 'center',
              display: 'flex',
              position: 'absolute',
              left: 'calc(50% - 25px)',
              top: '15%',
              background: 'linear-gradient(45deg, #F4F3E9 0%, #FFFFFF 100%)',
              border: '1px solid #F4F3E9',
              boxShadow:
                '-1px 1px 1px rgba(0, 0, 0, 0.09), -4px 4px 10px rgba(186, 186, 186, 0.25)',
              borderRadius: 100,
            }}
          >
            <img
              src="https://cdn.khulnasoft.com/api/v1/image/assets%2FYJIGb4i01jvw0SRdL5Bt%2F3d4768d27ae441ebaaba833eda935f21"
              style={{
                width: 25,
                height: 25,
                objectFit: 'contain',
                animation: 'spin 1.6s cubic-bezier(.85,.01,.29,1) infinite',
              }}
            />
          </div>
          <style>
            {`@keyframes spin {
              from {transform:rotate(0deg);}
              to {transform:rotate(360deg);}
            }`}
          </style>
        </>
      )}
      <div
        ref={ref}
        khulnasoft-liquid-block={khulnasoftBlock?.id}
        className="khulnasoft-liquid-block"
        dangerouslySetInnerHTML={{
          __html: html || cache[blockName + renderArgs] || '',
        }}
      />
    </>
  );
};

Khulnasoft.registerComponent(LiquidBlock, {
  name: 'Shopify:LiquidBlock',
  hideFromInsertMenu: true,
  inputs: [
    {
      name: 'templatePath',
      type: 'string',
      advanced: true,
    },
    {
      name: 'options',
      type: 'LiquidBlockOptions',
    },
  ],
});
