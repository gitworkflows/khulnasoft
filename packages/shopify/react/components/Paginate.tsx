import {
  KhulnasoftStore,
  withKhulnasoft,
  KhulnasoftStoreContext,
  KhulnasoftBlockComponent,
  KhulnasoftElement,
} from '@khulnasoft.com/react';
import * as React from 'react';

interface AssignBlockProps {
  expression?: string;
  limit?: string;
  khulnasoftState?: KhulnasoftStore;
  khulnasoftBlock?: KhulnasoftElement;
}

export function PaginateBlock(props: AssignBlockProps) {
  const defaultPaginationLimit = 15;
  const { expression, khulnasoftState, limit } = props;
  let paginate: any = {};

  if (
    expression &&
    khulnasoftState?.context?.shopify &&
    khulnasoftState.state.collection?.products?.length
  ) {
    const totalProducts = khulnasoftState.state.collection.products_count;
    const itemLimit =
      khulnasoftState.context.shopify.liquid.get(limit, khulnasoftState.state) || defaultPaginationLimit;

    let currentPage = 1;
    if (khulnasoftState.state.location.query.page) {
      // this is undefined, should probably add support to query params on location
      currentPage = parseInt(khulnasoftState.state.location.query.page, 10) || currentPage;
    } else if (typeof window !== 'undefined') {
      const urlParams = new URLSearchParams(window.location.search);
      currentPage = parseInt(urlParams.get('page') || '1', 10);
    }

    paginate = {
      current_page: currentPage,
      current_offset: currentPage * itemLimit,
      items: khulnasoftState.state.collection.products.length,
      parts: [],
      next: {},
      previous: {},
      page_size: itemLimit,
      pages: Math.ceil(totalProducts / itemLimit),
      expression: khulnasoftState.context.shopify.liquid.render(expression, khulnasoftState.state),
    };

    if (currentPage < paginate.pages) {
      paginate.next.url = `?page=${currentPage + 1}`;
    }

    if (currentPage > 1) {
      paginate.previous.url = `?page=${currentPage - 1}`;
    }
  }

  return (
    <KhulnasoftStoreContext.Consumer>
      {store => (
        <KhulnasoftStoreContext.Provider
          value={{
            ...store,
            state: {
              ...store.state,
              paginate,
            },
          }}
        >
          {props.khulnasoftBlock?.children?.map(item => (
            <KhulnasoftBlockComponent block={item} key={item.id} />
          ))}
        </KhulnasoftStoreContext.Provider>
      )}
    </KhulnasoftStoreContext.Consumer>
  );
}

withKhulnasoft(PaginateBlock, {
  name: 'Shopify:Paginate',
  hideFromInsertMenu: true,
  noWrap: true,
  inputs: [
    {
      name: 'expression',
      type: 'javascript',
    },
  ],
});
