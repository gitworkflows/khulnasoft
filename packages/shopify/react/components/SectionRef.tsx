import React, { useState, useEffect, useRef } from 'react';
import { Khulnasoft, KhulnasoftElement } from '@khulnasoft.com/react';
import { findAndRunScripts } from '../functions/find-and-run-scripts';

interface SectionRefProps {
  class?: string;
  section?: string;
  khulnasoftBlock?: KhulnasoftElement;
}
const cache: { [key: string]: string } = {};

const refs =
  Khulnasoft.isBrowser &&
  Array.from(document.querySelectorAll('[khulnasoft-shopify-section]')).reduce((memo, item) => {
    if (item.innerHTML) {
      memo[item.getAttribute('khulnasoft-shopify-section')!] = item;
    }
    return memo;
  }, {} as { [key: string]: Element });

export const SectionRef = (props: SectionRefProps) => {
  const [html, setHtml] = useState<string | null>(null);
  const sectionName = props.section?.split('/')[1]?.replace(/\.liquid$/, '')!;

  const ref = useRef<HTMLDivElement | null>(null);

  useEffect(() => {
    const blockId = props.khulnasoftBlock?.id;
    const node = blockId && refs && refs[blockId];
    if (!node) {
      if (cache[sectionName]) {
        return;
      }
      // Convert `sections/foo.liquid` -> `foo`
      if (!sectionName) {
        return;
      }
      if (cache[sectionName]) {
        setHtml(cache[sectionName]);
        return;
      }
      // This may look strange, but it is how Shopify's section rendering API
      // works https://shopify.dev/docs/themes/sections/section-rendering-api
      // TODO: also send ?preview_theme_id and save that as a cookie or wrap this API
      // specifically with /shopify/v1/data etc
      fetch(
        `https://cdn.khulnasoft.com/api/v1/proxy-api?url=${encodeURIComponent(
          `https://${
            location.host + location.pathname.replace('/apps/khulnasoft/preview', '')
          }?section_id=${sectionName}&${location.search.replace('?', '')}`
        )}`
      )
        .then(res => res.text())
        .then(text => {
          // to ensure we don't keep fetching content for empty sections, we need to insert something
          const safeText = text || '<span></span>';
          cache[sectionName] = safeText;
          setHtml(safeText);
          setTimeout(() => {
            if (ref.current) {
              findAndRunScripts(ref.current);
            }
          }, 10);
        });
    } else {
      if (node && ref.current) {
        ref.current.parentNode?.replaceChild(node, ref.current);
      }
    }
  }, [sectionName]);

  return !sectionName ? (
    <div style={{ padding: 50, textAlign: 'center' }}>
      Please choose a valid name for your section
    </div>
  ) : (
    <div
      ref={ref}
      khulnasoft-shopify-section={props.khulnasoftBlock?.id}
      className="khulnasoft-shopify-section"
      dangerouslySetInnerHTML={{ __html: html || cache[sectionName] || '' }}
    />
  );
};

Khulnasoft.registerComponent(SectionRef, {
  name: 'Shopify:SectionRef',
  friendlyName: 'Shopify Section',
  image:
    'https://cdn.khulnasoft.com/api/v1/image/assets%2FYJIGb4i01jvw0SRdL5Bt%2Ff3d80fca4afb49379d055b23e4c8b8c9',
  description: 'Include a Shopify section template',
  inputs: [
    {
      name: 'section',
      helperText: 'Full path to the section, e.g. sections/product.liquid',
      type: 'string',
      advanced: true,
    },
  ],
});
