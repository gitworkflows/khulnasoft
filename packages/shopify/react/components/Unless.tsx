import khulnasoft, {
  Khulnasoft,
  KhulnasoftBlockComponent,
  stringToFunction,
  withKhulnasoft,
} from '@khulnasoft.com/react';
import * as React from 'react';
import { UnlessBlockProps } from '../interfaces/component-props';

export class UnlessBlock extends React.Component<UnlessBlockProps> {
  private getMatchingBlocks() {
    const { expression, unlessBlocks, elseBlocks } = this.props;
    if (!expression) {
      return unlessBlocks;
    }
    const fn = stringToFunction(expression);
    const result = fn(
      this.props.khulnasoftState && this.props.khulnasoftState.state,
      null,
      this.props.khulnasoftBlock,
      khulnasoft,
      null,
      this.props.khulnasoftState!.update,
      Khulnasoft,
      this.props.khulnasoftState?.context
    );
    return result ? elseBlocks : unlessBlocks;
  }

  render() {
    const blocks = this.getMatchingBlocks();
    if (!blocks) {
      return null;
    }

    return blocks.map(block => <KhulnasoftBlockComponent key={block.id} block={block} />);
  }
}

const defaultBlock = {
  '@type': '@khulnasoft.com/sdk:Element',
  responsiveStyles: {
    large: {
      display: 'flex',
      flexDirection: 'column',
      alignItems: 'stretch',
      flexShrink: '0',
      position: 'relative',
      marginTop: '30px',
      textAlign: 'center',
      lineHeight: 'normal',
      height: 'auto',
    },
  },
  component: {
    name: 'Text',
    options: {
      text: '<p>Enter some text...</p>',
    },
  },
};

withKhulnasoft(UnlessBlock, {
  name: 'Shopify:Unless',
  hideFromInsertMenu: true,
  noWrap: true,
  inputs: [
    {
      name: 'expression',
      type: 'javascript',
    },
    {
      name: 'unlessBlocks',
      type: 'array',
      hideFromUI: true,
      defaultValue: [defaultBlock],
    },
    {
      name: 'elseBlocks',
      type: 'array',
      hideFromUI: true,
      defaultValue: [defaultBlock],
    },
  ],
});
