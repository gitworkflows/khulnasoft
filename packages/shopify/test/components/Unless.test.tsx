import { render } from '@testing-library/react';
import * as React from 'react';
import { UnlessBlock } from '../../react/components/Unless';
import { mockState, text, khulnasoftComponentIdRegex } from '../modules/helpers';
import * as reactTestRenderer from 'react-test-renderer';

describe('IfElse', () => {
  test('Basic Unless True', () => {
    const ref = render(
      <UnlessBlock
        expression="foo === 'bar'"
        unlessBlocks={[text('A')]}
        khulnasoftState={mockState({ foo: 'bar' })}
      />
    );

    expect(ref.queryByText('A')).toBeNull();
  });
  test('Basic Unless False', () => {
    const ref = render(
      <UnlessBlock
        expression="foo === 'bar'"
        unlessBlocks={[text('A')]}
        khulnasoftState={mockState({ foo: 'yo' })}
      />
    );

    expect(ref.queryByText('A')).toBeTruthy();
  });

  test('Basic Else False', () => {
    const ref = render(
      <UnlessBlock
        expression="foo === 'bar'"
        unlessBlocks={[text('A')]}
        elseBlocks={[text('B')]}
        khulnasoftState={mockState({ foo: 'bar' })}
      />
    );

    expect(ref.queryByText('A')).toBeNull();
    expect(ref.queryByText('B')).toBeTruthy();
  });

  test('Basic Else True', () => {
    const ref = render(
      <UnlessBlock
        expression="foo === 'bar'"
        unlessBlocks={[text('A')]}
        elseBlocks={[text('B')]}
        khulnasoftState={mockState({ foo: 'ya' })}
      />
    );

    expect(ref.queryByText('A')).toBeTruthy();
    expect(ref.queryByText('B')).toBeNull();
  });

  it('renders snapshot correctly', () => {
    const tree = JSON.stringify(
      reactTestRenderer
        .create(
          <UnlessBlock
            expression="foo === 'bar'"
            unlessBlocks={[text('A')]}
            elseBlocks={[text('B')]}
            khulnasoftState={mockState({ foo: 'bar' })}
          />
        )
        .toJSON()
    ).replace(khulnasoftComponentIdRegex, '');

    expect(tree).toMatchSnapshot();
  });
});
