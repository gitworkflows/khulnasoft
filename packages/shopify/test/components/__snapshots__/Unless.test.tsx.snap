// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IfElse renders snapshot correctly 1`] = `"{\\"type\\":\\"div\\",\\"props\\":{\\"style\\":{},\\"className\\":\\"khulnasoft-block  khulnasoft-has-component css-0\\",\\"khulnasoft-id\\":\\"\\"},\\"children\\":[{\\"type\\":\\"span\\",\\"props\\":{\\"className\\":\\"khulnasoft-text css-1qggkls\\",\\"dangerouslySetInnerHTML\\":{\\"__html\\":\\"B\\"}},\\"children\\":null}]}"`;
